File: leetcode/260. Single Number III.py
# Given an array of numbers nums, in which exactly two elements appear only once and all the other elements appear exactly twice. Find the two elements that appear only once.

# Example:

# Input:  [1,2,1,3,2,5]
# Output: [3,5]
# Note:

# The order of the result is not important. So in the above example, [5, 3] is also correct.
# Your algorithm should run in linear runtime complexity. Could you implement it using only constant space complexity?
class Solution:
    def singleNumber(self, nums):
        # 利用hash map
        s = {}
        if len(nums)<2:
            return -1
        if len(nums)==2:
            return nums
        for n in nums:
            if n not in s:
                s[n]=1
            else:
                s.pop(n)
        ans = []
        for key in s.keys():
            ans.append(key)
        return ans
s = Solution()
i = [1,2,1,3,2,5]
print(s.singleNumber(i))

File: leetcode/260. Single Number III.cpp
#include <iostream>
#include <unordered_set>
#include <algorithm>
#include <vector>
#include <string>
#include <numeric>
using namespace std;

class Solution {
public:
	vector<int> singleNumber(vector<int>& nums) {
		unordered_set<int> s;
		for (auto n:nums) {
			if (s.count(n) == 0)
				s.insert(n);
			else
				s.erase(n);
		}
		vector<int> a;
		for (auto n : nums) {
			a.push_back(n);
		}
		return a;
	}
};
int main() {
	vector<int> i = {1, 2, 3, 4, 0};
	Solution s;
	vector<int> ans = s.singleNumber(i);
	cout<<ans[0]<<ans[1]<< endl;
	system("pause");
	return 0;
}